doc/man/man4/itot.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/itot.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH ITOT 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.SH NAME
.B isot, itot
\- ISO Transport over TCP (ISOT/ITOT)
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.HP
.IB tcp_stream " = open(" \(lq/dev/tcp\(rq ", O_RDWR);
.HP
.BI ioctl( tcp_stream ", I_PUSH, " \(lqisot\(rq );
.sp
.HP
.IB tcp_stream " = open(" \(lq/dev/tcp\(rq ", O_RDWR);
.HP
.BI ioctl( tcp_stream ", I_PUSH, " \(lqitot\(rq );
.\"
.\"
.SH DESCRIPTION
.B ISOT
is an implementation of ISO Transport over TCP (ISOT) as described in RFC 983,
.[
rfc983
.]
superseded by RFC 1006.
.[
rfc1006
.]
.B ITOT
is an implementation of ISO Transport over TCP (ITOT) as described in RFC
2126.
.[
rfc2126
.]
Note that the former (RFC 1006) names the service
.BR ISOT ;
whereas the latter (RFC 2126) names the service
.BR ITOT .
.PP
Regardless of the naming differences, these RFCs describe the use of IS 8073
.[
ISO/IEC 8073 X.224
.]
ISO Transport Class 0 over TCP/IP (ISOT/ITOT) by using TCP/IP as a
Connection-oriented Network Service Provider instead of ISO CONS.
.[
ISO/IEC 8878 X.223
.]
.PP
They describe adaptation of TCP to an ISO CONS
.[
x223
.]
network service provider through the use of a 4 byte header used to delimit
NPDUs inside of the TCP stream.
.PP
RFC 983
.[
rfc983
.]
and RFC 1006
.[
rfc1006
.]
provided for ISO Transport Class 0 over TCP (or UDP under
protest) (ISOT) on port number 102.
(Note: both RFC 1006 and RFC 2126 place TP0 on port 102, and, therefore, RFC
2126 must be backward compatible with RFC 1006.)
RFC 1006
is a standard (STD 35).
.PP
RFC 2126
.[
rfc2126
.]
also provides for ISO Transport Class 0 over TCP (or UDP under protest) (ISOT)
on port number 102.  But, in addition to these services provided by RFC
1006/STD 35, RFC 2126 also provides for ISO Transport Class 2 over TCP (or UDP
under protest) (ITOT), based on RFC 1859,
.[
rfc1859
.]
using independent TCP/IP connections to port number 399, and also supports
both IPv4 and IPv6.
(Note: both RFC 1859 and RFC 2126 place TP2 on port 399, and, therefore, RFC
2126 must be backward compatible with RFC 1859.)
.\"
.\"
.SH USAGE
.PP
Not yet implemented, but here is a stab at explaining the design:
.PP
The following steps will be taken for use with the TP module:
.IP 1. \w'0.\(em'u
The user opens a
.BR tcp "(4) or " udp "(4)
Stream.
.IP 2.
The user pushes the
.BR isot "(4) or " itot "(4)
module.  The module, when pushed, determines the nature for the Stream below
it, whether TCP or UDP.
.RS
.PP
For ITOT Class 2 operation with expedited data, the user must open two
.BR tcp "(4) or " udp "(4)
Streams, or one
.BR tcp "(4) and one " udp "(4)
Stream, and link both under the
.BR itot (4)
multiplexer.  The linking is accomplished by opening a Stream on the
.RI \(lq /dev/itot \(rq
device and linking each Stream beneath the device Stream using the
.BR I_LINK (7)
STREAMS input-output control,
.BR streamio (7).
.RE
.IP 3.
The user pushes the
.BR tp0 "(4) or " tp2 "(4)
module.
.IP 4.
The user then treats the Stream as any other TP0 or TP2 Stream, with the
obvious restrictions on NSAP and QOS.
.PP
The following steps will be taken for use with the TP multiplexer:
.IP 1. \w'0.\(em'u
A TP connection manager opens a connection management Stream on the TP
multiplexing driver and listens for requests for the establishment of
network connections.
.IP 2.
A TP user opens an upper Stream for ISO Transport Class 0 or ISO Transport
Class 2.  The TP user then attempts to listen on the transport endpoint or
connect the transport endpoint.  If a suitable (determined by address format,
TSAP, and ownership) network connection is not available, it requests a
network connection from the connection manager.  (TP2 performs multiplexing,
so it is possible that many TP2 Streams could be opened and served by the same
network connection.)
.IP 3.
When the connection manager receives the request, it uses the address format,
TSAP and QOS parameters to determine whether an ISOT or ITOT network
connection is requested.
.IP 4.
If an ISOT or ITOT network connection is requested, the connection manager
opens a
.BR tcp "(4) or " udp "(4)
Stream.
.\"to the requested address and connects the Stream to the appropriate
.\"port number (102 for ISOT TP0 and 399 for ITOT TP2).
.IP 5.
The connection manager then pushes the
.BR isot " or " itot
module on the Stream.  This transforms the Stream into an ISO CONS provider
and presents the
.BR npi (7)
service interface at the upper boundary.
.IP 6.
The connection manager then links the ISOT or ITOT Stream under the TP
multiplexing driver and configures the Stream within the TP driver in response
to the network connection request.
.IP 7.
TP then proceeds as normal, unaware that TCP/IP is providing the ISO CONS
service.
.\"
.\"
.SH INTERFACE
.PP
The service interface for the ISOT and ITOT modules at the upper boundary is
the Network Provider Interface (NPI),
.[
npi revision 2.0.0 draft 2
.]
.BR npi (7).
The service interface for the ISOT and ITOT modules at the lower boundary is
the Transport Provider Interface (TPI),
.[
tpi revision 2.0.0 draft 2
.]
.BR tpi (7).
.\"
.\"
.SH ADDRESSES
.PP
Generic string addresses for ISOT and ITOT are described in RFC 1278.
.[
rfc1278
.]
.PP
OSI NSAP addresses for ISOT and ITOT are described for IPv4 in RFC 1277,
.[
rfc1277
.]
for IPv6 in RFC 1888
.[
rfc1888
.]
superseded by IPv4 and IPv6 in RFC 4548.
.[
rfc4548
.]
.\"
.\"
.SH OPTIONS
.PP
.PP
.B ISOT
and
.B ITOT
modules support the ISO CONS
options described in X.213
.[
x213
.]
and X.223.
.[
x223
.]
.\"
.\"
.SH PRIMITIVES
.PP
The
ISOT and ITOT modules mediate
.BR npi (7)
primitives from its upper interface (write queue) to 
.BR tpi (7)
primitives at its lower interface (read queue).
This section describes only the upper interface.  For a description of the
lower interface, see
.BR tcp (4)
and
.BR udp (4).
.SS N_BIND_REQ\fR(7), \fPN_BIND_ACK\fR(7)
.\"
.\"
.SH IOCTLS
.PP
ISOT and ITOT do not support any input-output controls of their own.
.\"
.\"
.SH NOTICES
.PP
These modules are not implemented.
.\"
.\"
.SH IMPLEMENTATION
.PP
These modules are not implemented.
.\"
.\"
.\".SH FILES
.\".PP
.\"
.\"
.SH MODULES
.PP
.TP
.B isot
the ISOT module is a
.I STREAMS
module named
.RI \(lq isot \(rq.
.TP
.B itot
the ITOT module is a
.I STREAMS
module named
.RI \(lq itot \(rq.
.\"
.\"
.SH DEVICES
.PP
.TP
.B /dev/tcp
The
.B tcp (4)
device opened onto which ISOT or ITOT is pushed.
.TP
.B /dev/udp
The
.B udp (4)
device opened onto which ISOT or ITOT is pushed.
.\"
.\"
.SH "SEE ALSO"
.PP
.IR ISO (4).
.\"
.\"
.SH BUGS
.PP
These modules are not implemented and so cannot have any bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
This implementation of ISOT and ITOT is specific to
.BR "@PACKAGE_TITLE@" .
.\"
.\"
.SH CONFORMANCE
.PP
ISOT conforms to RFC 1006.
.[
rfc1006
.]
ITOT confirms to RFC 2126.
.[
rfc2126
.]
.\"
.\"
.SH HISTORY
.PP
.BR ISOT " and " ITOT
are new for
.BR "@PACKAGE_TITLE@" .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/modules/itot.c
/*****************************************************************************

 @(#) File: src/modules/itot.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/modules/itot.c (" PACKAGE_ENVR ") " PACKAGE_DATE;


/*
 *  ISO Transport over TCP (ITOT)
 */

#include <sys/os7/compat.h>

/*
 *  These are for TPI definitions.
 */
#if defined HAVE_TIHDR_H
#   include <tidhr.h>
#else
#   include <sys/tihdr.h>
#endif

#include <sys/npi.h>
#include <net/sock.h>
#include <net/ip.h>

#define ITOT_DESCRIP	"RFC 2161 ISO Transport over TCP (ITOT) STREAMS Module"
#define ITOT_EXTRA	"Part of the OpenSS7 OSI Stack for Linux Fast-STREAMS"
#define ITOT_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define ITOT_REVISION	"OpenSS7 src/modules/itot.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define ITOT_DEVICE	"SVR 4.2 MP STREAMS ITOT Module for RFC 2126 (ITOT)"
#define ITOT_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define ITOT_LICENSE	"GPL"
#define ITOT_BANNER	ITOT_DESCRIP	"\n" \
			ITOT_EXTRA	"\n" \
			ITOT_COPYRIGHT	"\n" \
			ITOT_REVISION	"\n" \
			ITOT_DEVICE	"\n" \
			ITOT_CONTACT
#define ITOT_SPLASH	ITOT_DEVICE	" - " \
			ITOT_REVISION

#ifdef LINUX
MODULE_AUTHOR(ITOT_CONTACT);
MODULE_DESCRIPTION(ITOT_DESCRIP);
MODULE_SUPPORTED_DEVICE(ITOT_DEVICE);
#ifdef MODULE_LICENSE
MODULE_LICENSE(ITOT_LICENSE);
#endif				/* MODULE_LICENSE */
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-itot");
#endif				/* MODULE_ALIAS */
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif
#endif				/* LINUX */

#ifndef ITOT_MOD_NAME
#define ITOT_MOD_NAME		"itot"
#endif				/* ITOT_MOD_NAME */
#ifndef ITOT_MOD_ID
#define ITOT_MOD_ID		0
#endif				/* ITOT_MOD_ID */

/*
 *  STREAMS Definitions
 */

#define MOD_ID		ITOT_MOD_ID
#define MOD_NAME	ITOT_MOD_NAME
#ifdef MODULE
#define MOD_BANNER	ITOT_BANNER
#else				/* MODULE */
#define MOD_BANNER	ITOT_SPLASH
#endif				/* MODULE */

static struct module_info itot_minfo = {
	.mi_idnum = MOD_ID,		/* Module ID number */
	.mi_idname = MOD_NAME,		/* Module name */
	.mi_minpsz = 0,			/* Min packet size accepted */
	.mi_maxpsz = INFPSZ,		/* Max packet size accepted */
	.mi_hiwat = STRHIGH,		/* Hi water mark */
	.mi_lowat = STRLOW,		/* Lo water mark */
};

static struct module_stat itot_rstat __attribute__ ((aligned(SMP_CACHE_BYTES)));
static struct module_stat itot_wstat __attribute__ ((aligned(SMP_CACHE_BYTES)));

static streamscall int itot_open(queue_t *, dev_t *, int, int, cred_t *);
static streamscall int itot_close(queue_t *, int, cred_t *);

static streamscall int itot_rput(queue_t *, mblk_t *);
static streamscall int itot_wput(queue_t *, mblk_t *);

static struct qinit itot_rinit = {
	.qi_putp = itot_rput,		/* Read put (message from below) */
	.qi_qopen = &itot_open,		/* Each open */
	.qi_qclose = &itot_close,	/* Last close */
	.qi_minfo = &itot_minfo,	/* Module information */
	.qi_mstat = &itot_rstat,	/* Module statistics */
};

static struct qinit itot_winit = {
	.qi_putp = itot_wput,		/* Write put (message from above) */
	.qi_minfo = &itot_minfo,	/* Module information */
	.qi_mstat = &itot_wstat,	/* Module statistics */
};

struct streamtab itot_info = {
	.st_rdinit = &itot_rinit,
	.st_wrinit = &itot_winit,
};

/*
 *  PUTP and SRVP, put and service procedures.
 *
 *  ITOT does not provide service procedures.  This is primarily because the message handling across
 *  the module never amplifies message blocks (that is an allocation can never fail causing the
 *  incoming message to be placed onto the queue for later processing by the service procedure).
 */

/*
 *  Read-Side Processing
 */

/**
 * itot_r_proto - process M_(PC)PROTO message block on read side
 * @q: read queue
 * @mp: the message
 *
 * Note that much of this processing is stateless and pretty automatic.  When a request is generated
 * on the write side, we trasnform to an NPI message, pass it down and wait for the response.  Here
 * the response is most often simply translated into the corresponding NPI response.
 */
static streams_fastcall streams_inline __hot_get int
itot_r_proto(queue_t *q, mblk_t *mp)
{
	switch (*(t_scalar_t *) mp->b_rptr) {
	case T_CONN_IND:
	{
		/* Transform into N_CONN_IND. */
		struct prim {
			N_conn_ind_t prim;
			unsigned short family;
			char address[20];
		} n = { {
		N_CONN_IND,},};
		struct T_conn_ind *t = (typeof(t)) mp->b_rptr;
		struct sockaddr_in *sin =
		    (struct sockaddr_in *) ((unsigned char *) t + t->SRC_offset);
		int length;

		(void) sin;
		n.prim.PRIM_type = N_CONN_IND;
		n.prim.DEST_length = 0;
		n.prim.DEST_offset = 0;
		n.prim.SRC_length = t->SRC_length ? 20 : 0;
		n.prim.SRC_offset = sizeof(n);
		n.prim.SEQ_number = t->SEQ_number;
		n.prim.CONN_flags = 0;
		n.prim.QOS_length = 0;
		n.prim.QOS_offset = 0;

		if (n.prim.SRC_length) {
			n.family = AF_X25;
			n.address[0] = 0;
			n.address[1] = 0;
			n.address[2] = 0;

			n.address[3] = 0;
			n.address[4] = 0;
			n.address[5] = 0;
			n.address[6] = 0;

			/* the remainder are zero */

			length = sizeof(n);
		} else
			length = sizeof(n.prim);
		(void) length;
	}
	case T_CONN_CON:
	{
		/* Transform into N_CONN_CON. */
		N_conn_con_t n = { N_CONN_CON, };
		struct T_conn_con *t = (typeof(t)) mp->b_rptr;

		(void) n;
		(void) t;
	}
	case T_DISCON_IND:
	{
		/* Transform into N_DISCON_IND. */
		N_discon_ind_t n = { N_DISCON_IND, };
		struct T_discon_ind *t = (typeof(t)) mp->b_rptr;

		(void) n;
		(void) t;
	}
	case T_DATA_IND:
	{
		/* Transform into N_DATA_IND. */
		N_data_ind_t n = { N_DATA_IND, };

		(void) n;
	}
	case T_EXDATA_IND:
	{
		/* Transform into N_EXDATA_IND. */
		N_exdata_ind_t n = { N_EXDATA_IND, };

		(void) n;
	}
	case T_INFO_ACK:
	{
		/* Transform into N_INFO_ACK. */
		N_info_ack_t n = { N_INFO_ACK, };

		(void) n;
	}
	case T_BIND_ACK:
	{
		/* Transform into N_BIND_ACK. */
		N_bind_ack_t n = { N_BIND_ACK, };

		(void) n;
	}
	case T_ERROR_ACK:
	{
		/* Transform into N_ERROR_ACK. */
		N_error_ack_t n = { N_ERROR_ACK, };

		(void) n;
	}
	case T_OK_ACK:
	{
		/* Transform into N_OK_ACK. */
		N_ok_ack_t n = { N_OK_ACK, };

		(void) n;
	}
	case T_UNITDATA_IND:
	{
		/* Transform into N_UNITDATA_IND. */
		N_unitdata_ind_t n = { N_UNITDATA_IND, };

		(void) n;
	}
	case T_UDERROR_IND:
	{
		/* Transform into N_UDERROR_IND. */
		N_uderror_ind_t n = { N_UDERROR_IND, };

		(void) n;
	}
	case T_OPTMGMT_ACK:
	{
		/* Transform into N_OK_ACK. */
		N_ok_ack_t n = { N_OK_ACK, };

		(void) n;
	}
	case T_ORDREL_IND:
	{
		/* Transform into N_DISCON_IND. */
		N_discon_ind_t n = { N_DISCON_IND, };

		(void) n;
	}
	case T_OPTDATA_IND:
	{
		/* Transform into N_QOSDATA_IND, N_DATA_IND or N_EXDATA_IND. */
#ifdef N_QOSDATA_IND
		N_qosdata_ind_t n = { N_QOSDATA_IND, };

		(void) n;
#else
		int isnotexpedited = 0;

		if (isnotexpedited) {
			N_data_ind_t n = { N_DATA_IND, };

			(void) n;
		} else {
			N_exdata_ind_t n = { N_EXDATA_IND, };

			(void) n;
		}
#endif

	}
	case T_ADDR_ACK:
	{
		/* Transform into N_INFO_ACK (well, maybe just store the addresses for later
		   inclusion into an N_INFO_ACK. */
	}
	case T_CAPABILITY_ACK:
	{
		/* Transform into N_INFO_ACK. */
		N_info_ack_t n = { N_INFO_ACK, };

		(void) n;
	}

/*
 *  None of the following messages should appear at the read-side put procedure.  Only upward defined
 *  TPI messages should appear at the read-side put procedure.  Discard these.
 */
	case T_CONN_REQ:
	case T_CONN_RES:
	case T_DISCON_REQ:
	case T_DATA_REQ:
	case T_EXDATA_REQ:
	case T_INFO_REQ:
	case T_BIND_REQ:
	case T_UNBIND_REQ:
	case T_UNITDATA_REQ:
	case T_OPTMGMT_REQ:
	case T_ORDREL_REQ:
	case T_OPTDATA_REQ:
	case T_ADDR_REQ:
	case T_CAPABILITY_REQ:
		__swerr();
	default:
		freemsg(mp);
		return (0);
	}
}

static streamscall int
itot_r_flush(queue_t *q, mblk_t *mp)
{
	if (mp->b_rptr[0] & FLUSHR) {
		if (mp->b_rptr[0] & FLUSHBAND)
			flushband(q, mp->b_rptr[1], FLUSHDATA);
		else
			flushq(q, FLUSHDATA);
	}
	putnext(q, mp);
	return (0);
}

static streamscall int
itot_r_data(queue_t *q, mblk_t *mp)
{
	__swerr();
	return (EFAULT);
}

static streams_fastcall streams_noinline int
itot_r_msg(queue_t *q, mblk_t *mp)
{
	switch (mp->b_datap->db_type) {
	case M_PROTO:
	case M_PCPROTO:
		return itot_r_proto(q, mp);
	case M_DATA:
		return itot_r_data(q, mp);
	case M_FLUSH:
		return itot_r_flush(q, mp);
	}
	putnext(q, mp);
	return (0);
}

static streamscall __hot_get int
itot_rput(queue_t *q, mblk_t *mp)
{
	if (mp->b_datap->db_type == M_PROTO)
		return itot_r_proto(q, mp);
	return itot_r_msg(q, mp);
}

/*
 *  Write-Side Processing
 */

static streams_fastcall streams_inline __hot_put int
itot_w_proto(queue_t *q, mblk_t *mp)
{
	switch (*(np_long *) mp->b_rptr) {
	case N_CONN_REQ:
	case N_CONN_RES:
	case N_DISCON_REQ:
	case N_DATA_REQ:
	case N_EXDATA_REQ:
	case N_INFO_REQ:
	case N_BIND_REQ:
	case N_UNBIND_REQ:
	case N_UNITDATA_REQ:
	case N_OPTMGMT_REQ:

	case N_DATACK_REQ:
	case N_RESET_REQ:
	case N_RESET_RES:

	case N_CONN_IND:
	case N_CONN_CON:
	case N_DISCON_IND:
	case N_DATA_IND:
	case N_EXDATA_IND:
	case N_INFO_ACK:
	case N_BIND_ACK:
	case N_ERROR_ACK:
	case N_OK_ACK:
	case N_UNITDATA_IND:
	case N_UDERROR_IND:

	case N_DATACK_IND:
	case N_RESET_IND:
	case N_RESET_CON:

	default:
		freemsg(mp);
		return (0);
	}
}

static streamscall int
itot_w_flush(queue_t *q, mblk_t *mp)
{
	if (mp->b_rptr[0] & FLUSHW) {
		if (mp->b_rptr[0] & FLUSHBAND)
			flushband(q, mp->b_rptr[1], FLUSHDATA);
		else
			flushq(q, FLUSHDATA);
	}
	putnext(q, mp);
	return (0);
}

static streamscall int
itot_w_data(queue_t *q, mblk_t *mp)
{
	__swerr();
	return (EFAULT);
}

static streams_fastcall streams_noinline int
itot_w_msg(queue_t *q, mblk_t *mp)
{
	switch (mp->b_datap->db_type) {
	case M_PROTO:
	case M_PCPROTO:
		return itot_w_proto(q, mp);
	case M_DATA:
		return itot_w_data(q, mp);
	case M_FLUSH:
		return itot_w_flush(q, mp);
	}
	putnext(q, mp);
	return (0);
}

/**
 * itot_wput - write-side put procedure
 * @q: write queue
 * @mp: message to put
 *
 * Messages arriving at the write-side put procedure are messages from above that are NPI formatted
 * messages.  For all intents and purposes, the module above us should believe that we are any X.213
 * based ISO CONS provider.
 */
static streamscall __hot_put int
itot_wput(queue_t *q, mblk_t *mp)
{
	int type = mp->b_datap->db_type;

	/* Fast path. */
	if (type == M_PROTO)
		return itot_w_proto(q, mp);
	return itot_w_msg(q, mp);
}

static streamscall int
itot_open(queue_t *q, dev_t *devp, int oflags, int sflag, cred_t *crp)
{
	return (ENXIO);
}

static streamscall int
itot_close(queue_t *q, int oflags, cred_t *crp)
{
	return (ENXIO);
}
